### ISSUE TITLE
`TypeError: undefined is not a function (evaluating '$(".thumbnail").fancybox')`

### ISSUE THREAD

**Issue by XhmikosR** [giakki/uncss#220]
So, I keep getting these TypeErrors with most of my projects. This doesn't happen with 0.12.1; it happens with master too.

See [this](https://travis-ci.org/mpc-hc/mpc-hc.org/builds/112209168) for example.

/CC @giakki @mikelambert

**Comment by XhmikosR** [comment]
So, any ideas, guys? Is this because of https://github.com/giakki/uncss/commit/62145b351fa9eedccdd8eeb5b8d5f14db72e4100 and it just didn't show the errors with 0.12.1?

**Comment by mikelambert** [comment]
Because you're running uncss against all your html files.

But some of your html files are simple includes, and are not standalone html files that can execute properly on their own:
https://github.com/mpc-hc/mpc-hc.org/blob/24e162148cfd9a842f3e28b78fb385f0256b717c/source/_includes/fancybox.html

**Comment by XhmikosR** [comment]
Nope, that is the source obviously. It's just a Jekyll include.

**Comment by mikelambert** [comment]
You sure that include is not being matched by "<%= dirs.dest %>/**/*.html"?
The error looks exactly like what you'd see if Uncss was processing an
include html file that depended on Jquery, but did not import the Jquery
library explicitly...
2016年3月6日(日) 19:30 XhmikosR <notifications@github.com>:

**Comment by XhmikosR** [comment]
Yeah. The dest dir has 52 HTML files and jQuery is loaded in all of them.

**Comment by XhmikosR** [comment]
See https://travis-ci.org/mpc-hc/mpc-hc.org/builds/114187093 and https://travis-ci.org/mpc-hc/mpc-hc.org/builds/114187157. The first is with grunt-uncss 0.4.4, which is using uncss 0.12.1, the latter is with uncss from master, but happens with 0.13.0 too.

**Comment by mikelambert** [comment]
No idea then.

You can of course try to isolate down to a single small html that repros
the issue, which is what ideally we'd request from someone asking for help
on these issues. ;)

But there's a year's worth of changes between 0.12 and 0.13? You could do a
binary-search on the changes over the past year, to figure out which change
triggers the error. Though that may or may not actually help fix the issue,
and you might still need that small repro case to see what input data is
causing issues.

2016年3月7日(月) 16:55 XhmikosR <notifications@github.com>:

**Comment by XhmikosR** [comment]
I know the procedure, and I have tried to isolate this but it's hard. That is why I decided to make an issue in case someone else has better luck.

The thing is I get such parsing errors in other projects too.

**Comment by XhmikosR** [comment]
What I can say for sure is that if i remove https://github.com/mpc-hc/mpc-hc.org/blob/master/source/_includes/fancybox.html the error goes away, i.e. I don't see any other parsing errors.

I tried even doing this

```html
<script>
    $(document).ready(function () {
        $(".thumbnail").fancybox({});
    });
</script>
```

still the same error. It's like phantomjs doesn't like fancybox  :/

**Comment by XhmikosR** [comment]
@giakki: so, do you have any ideas what's going on here? I looked at https://github.com/giakki/uncss/compare/0.12.1...0.13.0 and the only change I can think of is https://github.com/giakki/uncss/commit/b44948b20de34368a8772e0c954e24015bdc45c4 or https://github.com/giakki/uncss/commit/62145b351fa9eedccdd8eeb5b8d5f14db72e4100

I have tried everything I could think of. The type error is thrown with fancybox regardless that jQuery **is** loaded and the call to it is fine.

**Comment by XhmikosR** [comment]
Something else I noticed, is that if I include the code in pack.js then I don't get the error either.

**Comment by jegtnes** [comment]
I have the same issue as @XhmikosR. Not with Fancybox, but with a third-party Slideshare dependency on some of our pages (Slideshare) logging errors that don't appear on the live site.

This doesn't actually create any problems with the UnCSS output as the Slideshare assets aren't run through UnCSS. The only issue it's causing is the extra noise in our CI/build logs. That being said, I was just wondering if is there's a way to optionally silence or filter the output of Phantom's `stderr`s—or any plans to do so?

For what it's worth (probably not much), the errors I'm getting are below:
```
TypeError: 'undefined' is not a function (evaluating 'this.displayProgressTooltip.bind(this)')

  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:1
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:1 in SSPlayerController
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:1 in SSPlayer
  http://www.slideshare.net/slideshow/embed_code/8661735:29
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:1 in s
TypeError: 'undefined' is not a function (evaluating 'this.displayProgressTooltip.bind(this)')

  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:1
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:3
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:1 in SSPlayerController
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:4
  http://public.slidesharecdn.com/b/ss_foundation/combined_player_presentation.js?77549234d8:1 in SSPlayer
  http://www.slideshare.net/slideshow/embed_code/13766370:21
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:2
  http://public.slidesharecdn.com/b/ss_foundation/combined_jquery.js?7ce27c46c9:1 in s
```

**Comment by XhmikosR** [comment]
@jegtnes: just to confirm, can you try 0.12.1? That version shouldn't give the error, or at least it doesn't for me.

**Comment by jonmhutch7** [comment]
I'm having the same issues with the Mapbox API. Like @jegtnes says, it doesn't affect the actual CSS output, its just super freaking annoying in the console.

**Comment by jegtnes** [comment]
@XhmikosR Sorry, missed this message. When originally reported I tried this in both 0.12.1 and 0.13 with the same noisy output.

**Comment by XhmikosR** [comment]
That's weird because in my case 0.12 works fine.

**Comment by XhmikosR** [comment]
@mikelambert:

```html
<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>uncss TypeError demo</title>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.1.1/normalize.min.css">
    <style>
        .content {
            margin: 20px;
        }

        .red {
            color: red;
        }

        .green {
            color: green;
        }
    </style>
</head>

<body>
    <div class="content">
        <span class="green">This text should be green and change to red after page load.</span>
    </div>
    <!-- This works fine -->
    <!--<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>-->
    <!-- This throws the following with v0.13.0:
            ReferenceError: Can't find variable: $

              local:37

        With the master version:
            ReferenceError: Can't find variable: $

              local:35 in global code
    -->
    <script src="jquery-2.2.4.min.js"></script>
    <script>
        $(function() {
            $(".green").removeClass("green").addClass("red");
        });
    </script>
</body>
</html>

```

You need a local jquery.js; if you load jquery from a CDN (see the comment) it works.

Tested with v0.13.0.

Also, I see the output CSS has a sourcemap reference which is wrong. What's also wrong is that I see selectors that are not needed like `::-webkit-input-placeholder` and `::-webkit-file-upload-button`.

```css
/*** uncss> filename: https://cdnjs.cloudflare.com/ajax/libs/normalize/4.1.1/normalize.min.css ***/html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit-text-size-adjust:100%}body{margin:0}::-webkit-input-placeholder{color:inherit;opacity:.54}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}/*# sourceMappingURL=normalize.min.css.map */
```

**Comment by XhmikosR** [comment]
After digging into the TypeError, I found that the cause is https://github.com/giakki/uncss/blob/master/src/phantom.js#L85

```js
...
        } else if (ignoreRequests.test(url)) {
            networkRequest.abort();
}
...
```

Not sure what the side effects can be, but the error goes away with `cancel()`.

See #242

**Comment by XhmikosR** [comment]
So, was anyone able to find a solution about this?

**Comment by mikelambert** [comment]
So what's interesting here, is that the `abort`-vs-`cancel` code should only run if `ignoreRequests` check passes. And ignoreRequests is `/\.css$|\.png$|\.gif$|\.jpg$|\.jpeg$|fonts\.googleapis/`. So I don't see why that would matter for the jquery js file at all...

However, I see that you are using a file-relative URL, and the file is likely loaded locally off disk. So I wouldn't be surprised if the `file://` URL checks are applying here, and according to phantom.js, it might try to rewrite them as `http://` urls (which wouldn't work, and fail to load the jquery file). And a failed jquery load might result in the error you see?

But that's just a hypothesis here...I forget how uncss/phantom deal when it fails to load a javascript file referenced from html.

**Comment by XhmikosR** [comment]
@mikelambert: I think you are on the right track. While trying to debug the issue, I see uncss going into

```js
        } else if (ignoreRequests.test(url)) {
            networkRequest.abort();
}
```

**Comment by mikelambert** [comment]
In your debugging, can you print `ignoreRequests` and `url`, and see what is actually running when your code attempts to run `networkRequest.abort()`? Is it the jquery file, or another file, that is triggering the `abort/cancel` for your proposed fix?

**Comment by XhmikosR** [comment]
Using the sample from https://github.com/giakki/uncss/issues/220#issuecomment-223927766, I get:

```
C:\Users\xmr\Desktop>node uncss\bin\uncss index.html
>>> ignoreRequests /\.css$|\.png$|\.gif$|\.jpg$|\.jpeg$|fonts\.googleapis/
>>> originalUrl https://cdnjs.cloudflare.com/ajax/libs/normalize/4.1.1/normalize
.min.css
>>> url https://cdnjs.cloudflare.com/ajax/libs/normalize/4.1.1/normalize.min.css

ReferenceError: Can't find variable: $

  local:40 in global code
/*** uncss> filename: https://cdnjs.cloudflare.com/ajax/libs/normalize/4.1.1/nor
malize.min.css ***/html{font-family:sans-serif;-ms-text-size-adjust:100%;-webkit
-text-size-adjust:100%}body{margin:0}::-webkit-input-placeholder{color:inherit;o
pacity:.54}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}/
*# sourceMappingURL=normalize.min.css.map */

```

**Comment by XhmikosR** [comment]
Hmm, is `ignoredExtensions` just wrong since it ignores `*.css` files?

**Comment by XhmikosR** [comment]
Should be fixed with 0.15.
